source/opt/aggressive_dead_code_elim_pass.cpp

```diff
@@ -0,0 +1,210 @@
+// Copyright (c) 2017 The Khronos Group Inc.
+// Copyright (c) 2017 Valve Corporation
+// Copyright (c) 2017 LunarG Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "aggressive_dead_code_elim_pass.h"
+
+#include "iterator.h"
+
+static const int kSpvEntryPointFunctionId = 1;
+static const int kSpvStorePtrId = 0;
+static const int kSpvLoadPtrId = 0;
+static const int kSpvAccessChainPtrId = 0;
+static const int kSpvTypePointerStorageClass = 0;
+static const int kSpvCopyObjectOperand = 0;
+static const int kSpvNameTargetId = 0;
+
+namespace spvtools {
+namespace opt {
+
+bool AggressiveDCEPass::IsNonPtrAccessChain(const SpvOp opcode) const {
+  return opcode == SpvOpAccessChain || opcode == SpvOpInBoundsAccessChain;
+}
+
+ir::Instruction* AggressiveDCEPass::GetPtr(
+      ir::Instruction* ip, uint32_t* varId) {
+  *varId = ip->GetSingleWordInOperand(
+      ip->opcode() == SpvOpStore ?  kSpvStorePtrId : kSpvLoadPtrId);
+  ir::Instruction* ptrInst = def_use_mgr_->GetDef(*varId);
+  ir::Instruction* varInst = ptrInst;
+  while (varInst->opcode() != SpvOpVariable) {
+    if (IsNonPtrAccessChain(varInst->opcode())) {
+      *varId = varInst->GetSingleWordInOperand(kSpvAccessChainPtrId);
+    }
+    else {
+      assert(varInst->opcode() == SpvOpCopyObject);
+      *varId = varInst->GetSingleWordInOperand(kSpvCopyObjectOperand);
+    }
+    varInst = def_use_mgr_->GetDef(*varId);
+  }
+  return ptrInst;
+}
+
+bool AggressiveDCEPass::IsLocalVar(uint32_t varId) {
+  const ir::Instruction* varInst = def_use_mgr_->GetDef(varId);
+  assert(varInst->opcode() == SpvOpVariable);
+  const uint32_t varTypeId = varInst->type_id();
+  const ir::Instruction* varTypeInst = def_use_mgr_->GetDef(varTypeId);
+  return varTypeInst->GetSingleWordInOperand(kSpvTypePointerStorageClass) ==
+      SpvStorageClassFunction;
+}
+
+void AggressiveDCEPass::AddStores(uint32_t ptrId) {
+  analysis::UseList* uses = def_use_mgr_->GetUses(ptrId);
+  if (uses == nullptr)
+    return;
+  for (auto u : *uses) {
+    SpvOp op = u.inst->opcode();
+    if (op == SpvOpStore)
+      worklist_.push(u.inst);
+    else if (op != SpvOpLoad)
+      AddStores(u.inst->result_id());
```

Comment:
I wonder if we could get into an infinite loop.
For example, with SPV_KHR_variable_pointers, a pointer can be the result of an OpPhi.
So if you stick that in a loop, then you can get a cycle where a pointer could trace back to itself.

E.g.


```
%entry = OpLabel
OpBranch %loop

%loop =OpLabel
%ptr = OpPhi %ptr_type %ptr %loop  %loop %base_ptr %entry
...
OpBranchConditional %loop
```

Suggest fixing this with a container of visited ptrIds.